[46] Residents can locate problem

leeky opened the pull request:
A resident can visit `/address_search/` and enter their postcode. If it is in the correct format, they are shown a list of matching addresses.

If they have JavaScript enabled, the submit buttons are hidden until they select an address from the list. 

This is a first attempt at this, and will need revisiting in the future (clicking back button, shows radio button previously selected, but the submit button is still hidden, for example). I'm suspecting this should be some standalone JS tests (perhaps using `jasmine-rails`), as this behaviour should be decoupled from the app.

NB: Currently this entirely returns canned responses in place of the API, which is still being built.

leeky committed:
(feature) Residents can locate the problem

- Search for addresses by postcode
- Select an address and use it on the next page

Approach:
- use i18n instead of hardcoded text
- use SimpleForm with named routes
- for this inital commit, hardcode data into the templates: we'll push
it down into the controller and beyond in the next step

dgmstuart committed:
(feature) addresses are returned by a query object

- introduce a new type of object: queries, responsible for fetching
data. We expect to have a number of these
- for now have this just return the hardcoded data

NB: This commit originally added `app/queries` to the eager loaded path,
which is not actually needed in Rails 5 and above:

> config.eager_load_paths accepts an array of paths from which Rails
> will eager load on boot if cache classes is enabled. Defaults to every
> folder in the app directory of the application.

http://edgeguides.rubyonrails.org/configuring.html#rails-general-configuration

dgmstuart committed:
(refactor) AddressFinder returns objects

This is perhaps a trade-off:
- Address finder isn't building a structure specifically for views (an
array of arrays)
- The view needs to know the names of the data attributes it's
displaying
  - On the other hand this feels easier to understand than :first/:last

dgmstuart committed:
(feature) Delegate fetching properties to an API

This forces us to drive out how we're sending the search data from the
form to the API.

- For now let's treat HackneyAPI as the boundary of the system and have
it be a class that has methods which return JSON (hence the string
keys). Later we might push this down further to a base JSON API (which
will probably just be Faraday)
- The AddressSearch code is maybe YAGNI at the moment (we could just
call 'postcode' on the form object), but I'm keen to separate out early
the difference between the data which we _validate_ (and potentially
show back to the user) from the data which we _sanitise_ (and send to
the query) since we've been bitten by this in the past
- I decided to name the result objects 'Property': in this context we're
using the addresses, but the data we're getting is property data.
- Note the use of 'spy' in one of the tests: We haven't used this much
but I like it for this sort of assertion: https://relishapp.com/rspec/rspec-mocks/docs/basics/spies

leeky committed:
(feature) Address search delegates to a JSON api

Started to build out the HackneyApi wrapper class that will talk to
Hackney's API.

The search results are hidden until we have performed a search.

We have assumed that URL encoding will be handled by the HTTP library.

Since this page will be shown initially before the user has done a
search, and then will show search results when they have done so, we
can't use strong parameters `require` here.

leeky committed:
(feature) SelectedAnswerStore stores answers in the session

This approach of storing answers in the session was taken directly from
https://github.com/alphagov/verify-frontend. It provides a lightweight
wrapper around Rails' session store.

Pass a hash of values for each page - probably corresponding to the form
fields on that page.

Note: because session stores can't in general handle symbols, the data
can only be accessed using string keys

leeky committed:
(feature) Pass property_reference around, but show full address

The radio button on the address search results page passes the
property_reference around, but we need to show the full address
to the user, and also store this information in the session.

In order to do this, I've expanded the JsonApi endpoint to fake out
two separate API routes. This allows different results to be
returned for requests to get properties by postcode vs. get details
of an individual property.

- Using the SelectedAnswerStore to store answers and use that to fetch
the individual property from the API

dgmstuart committed:
(feature) display a message when no results found

I'm assuming that we'll want the same 'address_search_results' div
whether or not there are actually results. That might not be right.

dgmstuart committed:
(feature) Postcode is validated on address search

- Validate presence
- Validate UK postcode using a regexp I found in a Gist - this might not
  be 100% what we need, but is easy to switch out for something else later
  - The regexp is very long so let's exclude it from rubocop
  - Note that it rejects postcodes with random spaces in them. If we
    want to fix that later we could try stripping *all* whitespace
    before validating.
- Note that we need to change the logic for when we show results:
  1. Don't show results or validation errors on first visit
  2. If there are validation errors, show them and no results
  3. Otherwise show results and no errors
- We should consider hiding the search box when we have results as per
  https://www.gov.uk/service-manual/design/addresses#using-an-address-lookup
  - this would get around having the conditionals in the view and
    controller
- Note that currently we're showing both inline and full error messages:
  my preference would be to just use the latter for these single field
  forms, but I'm open to ideas about this
- Briefly experimented with Shoulda Matchers, decided against:
  - it requires depending on a branch on github for rails 5
  - it uses 'try' so needs to be in rails_helper, even if the model
    doesn't depend on Rails
- The specs here are quite comprehensive - since we're using a
  third party regexp, I wanted to make sure we covered all the
  possibilities
- I feel like testing the validations in the model (rather than in
  isolation) is fine here: testing a module is fiddly and I don't
  anticipate that we'll reuse this validation.
- Tell Simplecov that Validators exist so that it groups them in the
  output

dgmstuart committed:
(feature): normalise postcodes in form output

Ensure postcodes are in a standard format before we send them to the
query object.

Is this YAGNI?

Should this be the responsibility of a different object?

I'm conflicted on whether this is better to do only on output, or when
presenting back to the user as well. If we did it up front we'd have to
override the getter or setter, and we'd have to handle nil.

leeky committed:
(spec) Test no postcode entered vs invalid postcode

leeky committed:
(feature) First pass at locking postcode entry after search

As per the GDS design pattern for postcode entry, I've started to split
the postcode search into two parts:

 - The first only asks to the user for a postcode, and handles
   validation of that postcode
 - The second displays a list of addresses to choose from, and gives the
   user the option of changing the postcode (by clicking a button that
   returns them to the first screen).

This means that there's only one form interaction on a page.

Source: https://www.gov.uk/service-manual/design/addresses#using-an-address-lookup

leeky committed:
(refactor) Address search form submits via PUT

Now the `#show` action shows the initial postcode search form (and any
validation errors) only. The form submits to the `#update` action which
shows the search results to the user to choose from.

leeky committed:
(chore) Change button label is i18n'd

leeky committed:
(feature) Redirect back to search form if no address selected

leeky committed:
(chore) Install Poltergeist so we can test pages with JS

NB: I'm not currently adding `database_cleaner` as we're not persisting
any data at present.

leeky committed:
(feature) Submit button is hidden until a radio button chosen

This is a fairly naive approach, but good enough for now. It:

 - Loops over every form on a page, looking for those that contain
   radio buttons.
 - When it finds one, it hides any buttons on that form
 - Finally, an onClick handler is added to the radio buttons that
   shows the form's button.

dgmstuart commented on spec/features/resident_can_locate_problem_spec.rb:
```diff
@@ -60,7 +60,16 @@
+  scenario 'when no postcode is entered' do
+    visit '/address_search/'
+
+    fill_in :address_search_postcode, with: '  '
```
I'd normally only test _one_ validation failure in a feature spec, the idea being that that is sufficient to test that everything integrates correctly. 

No harm in more - the tradeoff is with speed.

dgmstuart commented on spec/models/address_search_spec.rb:
```diff
@@ -24,6 +24,10 @@
       expect(AddressSearch.new(params).data).to eq(postcode: 'N16 8QR')
     end
+
+    it 'returns the search data even if no params were provided' do
```
I can't work out why this is - I've checked and nothing seems to break if it's not true.

Is my understanding correct?: if no params are provided then the form is invalid, and if the form is invalid then the caller shouldn't be asking it for data.

If that's the case then would it be better to raise an exception in this scenario?

dgmstuart replied:
Was removed

dgmstuart commented on app/controllers/address_searches_controller.rb:
```diff
@@ -1,11 +1,15 @@
 class AddressSearchesController < ApplicationController
   def show
-    @address_search = AddressSearch.new(address_search_params[:address_search])
```
Is this still "showing" an address search, or creating a new one?

Would it be better for these two actions to be the normal pairs? new/create (or edit/update?)

dgmstuart replied:
(changed)

dgmstuart commented on app/controllers/address_searches_controller.rb:
```diff
@@ -1,11 +1,15 @@
+    unless @address_search.valid?
+      render :show
+      return
+    end
```
How would it look if the happy case redirected to an "index" action on the Addresses controller instead?

dgmstuart replied:
(discussed - that would mean having to handle the case where that page is viewed without a postcode having been selected)

dgmstuart commented on app/views/address_searches/update.html.haml:
```diff
@@ -0,0 +1,16 @@
+
+= simple_form_for @address_search, url: address_search_path, method: :get do |f|
+  = @address_search.postcode
+  = f.button :submit, 'Change'
```
Since this doesn't have any fields, could it be a link instead of a form?

dgmstuart replied:
(changed)

dgmstuart replied:
Done

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -1,13 +1,16 @@
+    if property_reference.empty?
+      return redirect_back(
+        fallback_location: address_search_path,
+        alert: 'Please choose your address to continue'
```
Should this alert string go in the locales?

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -1,13 +1,16 @@
 class AddressesController < ApplicationController
   def update
-    selected_answers_store = SelectedAnswerStore.new(session)
```
This commit says "TODO" - does that mean you intended to squash it in with another commit?

leeky replied:
I've removed it and added an additional card in the backlog.

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -17,4 +20,8 @@ def update
+
+  def property_reference
+    update_params[:property_reference]
+  end
```
Is there a benefit to extracting this function? The tradeoff is that it hides away useful information at the bottom of the file and as a controller grows it can become unclear which actions the function is used by.

I know it's a common pattern from TVH, but I'm not sure it's a good one!

leeky replied:
I agree with you. The change was only necessary to keep RuboCop happy. 

We should consider allowing more lines per methods, perhaps?

dgmstuart commented on spec/features/resident_can_locate_problem_spec.rb:
```diff
@@ -90,4 +90,24 @@
+
+    fake_api = instance_double(JsonApi)
+    allow(fake_api).to receive(:get).with('properties?postcode=N1 6NU').and_return([matching_property])
+    allow(fake_api).to receive(:get).with('properties/').and_return({})
```
Is this line required? The test seems to pass without it.

leeky replied:
Good spot. I think this was left over from some earlier testing :+1:

dgmstuart commented on app/controllers/addresses_controller.rb:
```diff
@@ -1,13 +1,16 @@
-    selected_answers_store = SelectedAnswerStore.new(session)
-
-    property_reference = update_params[:property_reference]
+    if property_reference.empty?
```
Would this make sense as a validation on a form object - or on the Address model?

dgmstuart replied:
(maybe, but this works fine)
This should be `blank?` though

dgmstuart commented on app/assets/javascripts/form.js:
```diff
@@ -0,0 +1,18 @@
+
+    // If the form contains radio buttons, hide the submit button
+    if ($(this).find('input[type=radio]').length > 0) {
+      $(this).find('input[type=submit]').hide();
```
Do we also need to account for `input[type=button]`, or will we just avoid using that?

dgmstuart commented on spec/features/resident_can_locate_problem_spec.rb:
```diff
@@ -91,7 +91,32 @@
+
+    fake_api = instance_double(JsonApi)
+    allow(fake_api).to receive(:get).with('properties?postcode=N1 6NU').and_return([matching_property])
+    allow(fake_api).to receive(:get).with('properties/').and_return({})
```
As per previous comment - I think this line might not be required?

dgmstuart replied:
Done

dgmstuart commented on spec/features/resident_can_locate_problem_spec.rb:
```diff
@@ -91,7 +91,32 @@
+      choose 'Flat 1, 8 Hoxton Square, N1 6NU'
+    end
+
+    expect(page).to have_button t('helpers.submit.address.create')
```
Nice test 👍

dgmstuart reviewed (changes requested):
The approach looks great and it seems to work like a charm. I have a bunch of questions about the code.

Not sure why all my comments showed up as being on old commits - maybe some rebasing went on?

Also I'd like to squash in the FIX commits I made previously - or reword them if you think they belong as first class citizens.

leeky committed:
(fix) Use correct HTTP verbs when searching

 - Rename `#show` to `#new`
 - Rename `#update` to `#create`

This has a side-effect of simplifying the forms, which no longer need an
explict `method` parameter.

dgmstuart commented:
@leeky 

As discussed, the following spec will fail:
```ruby 
it 'adds a space in the middle' do
  params = { postcode: 'N168QR' }
  expect(AddressSearch.new(params).data).to eq(postcode: 'N16 8QR')
end
```

I'm now feeling like we should have two separate objects: 

- a Form object which is responsible for validation 
- a Formatter object which will need to be used (by coincidence) to present data to both the view and the query. This should expect a valid postcode and should raise some error if the data it's given can't be formatted as a postcode.

dgmstuart commented:
rebasing

dgmstuart committed:
(feature) Postcodes get a space at the right place

Negative indexing is odd: to put a space 3 from the end the index is
-4, since the last character is at index -1

dgmstuart committed:
(feature) Addresses controller handles nil refs

If the property_reference is nil, then something has gone quite wrong -
someone has posted to this action with invalid parameters. Still, best
to catch that rather than let it go through to the API call.

This doesn't have a test - I don't think an integration spec is
justifiable (?). To unit test this we would extract a form object, but
we agreed not to do that for now.

dgmstuart committed:
(refactor) move address errors into locales

Chose 'addresses' rather than 'address', since we're talking about a
list of addresses here, but I'm not attached to it.

leeky commented on app/models/address_search.rb:
```diff
@@ -12,7 +12,7 @@ def data
 
   def normalised_postcode
-    postcode.strip.gsub(/\s+/, ' ').upcase
+    postcode.strip.gsub(/\s+/, '').insert(-4, ' ').upcase
```
Didn't know about `insert` :+1:
